File: rideshare-dashboard/src/scripts/check-i18n-parity.ts
import { translations } from "../i18n/translations.ts";

type TranslationRecord = Record<string, string>;

function collectKeys(record: TranslationRecord): Set<string> {
  return new Set(Object.keys(record));
}

function findEmptyValues(
  record: TranslationRecord,
  locale: string,
): string[] {
  return Object.entries(record)
    .filter(([, value]) => value.trim().length === 0)
    .map(([key]) => `${locale}.${key}`);
}

const enKeys = collectKeys(translations.en);
const arKeys = collectKeys(translations.ar);

const onlyEn = [...enKeys].filter((key) => !arKeys.has(key)).sort();
const onlyAr = [...arKeys].filter((key) => !enKeys.has(key)).sort();
const emptyValues = [
  ...findEmptyValues(translations.en, "en"),
  ...findEmptyValues(translations.ar, "ar"),
];

let failed = false;

if (onlyEn.length > 0) {
  failed = true;
  console.error("Keys in en missing from ar:");
  for (const key of onlyEn) {
    console.error(`  - ${key}`);
  }
}

if (onlyAr.length > 0) {
  failed = true;
  console.error("Keys in ar missing from en:");
  for (const key of onlyAr) {
    console.error(`  - ${key}`);
  }
}

if (emptyValues.length > 0) {
  failed = true;
  console.error("Empty translation values:");
  for (const key of emptyValues) {
    console.error(`  - ${key}`);
  }
}

if (failed) {
  process.exit(1);
}

console.log(
  `i18n parity OK: ${enKeys.size} keys matched between en and ar`,
);

File: rideshare/tool/check_i18n_parity.dart
// ignore_for_file: avoid_print
import 'dart:convert';
import 'dart:io';

/// Verifies app_en.arb and app_ar.arb have identical key sets and non-empty values.
void main() {
  final enFile = File('lib/l10n/app_en.arb');
  final arFile = File('lib/l10n/app_ar.arb');

  if (!enFile.existsSync() || !arFile.existsSync()) {
    stderr.writeln('ARB files not found under lib/l10n/');
    exit(1);
  }

  final enKeys = _messageKeys(enFile);
  final arKeys = _messageKeys(arFile);

  final onlyEn = enKeys.difference(arKeys).toList()..sort();
  final onlyAr = arKeys.difference(enKeys).toList()..sort();

  var failed = false;

  if (onlyEn.isNotEmpty) {
    failed = true;
    stderr.writeln('Keys in app_en.arb missing from app_ar.arb:');
    for (final key in onlyEn) {
      stderr.writeln('  - $key');
    }
  }

  if (onlyAr.isNotEmpty) {
    failed = true;
    stderr.writeln('Keys in app_ar.arb missing from app_en.arb:');
    for (final key in onlyAr) {
      stderr.writeln('  - $key');
    }
  }

  for (final key in enKeys.intersection(arKeys)) {
    final enValue = _messageValue(enFile, key);
    final arValue = _messageValue(arFile, key);
    if (enValue.trim().isEmpty) {
      failed = true;
      stderr.writeln('Empty EN value for key: $key');
    }
    if (arValue.trim().isEmpty) {
      failed = true;
      stderr.writeln('Empty AR value for key: $key');
    }
  }

  if (failed) {
    exit(1);
  }

  print(
    'i18n parity OK: ${enKeys.length} keys matched between app_en.arb and app_ar.arb',
  );
}

Set<String> _messageKeys(File file) {
  final json = jsonDecode(file.readAsStringSync()) as Map<String, dynamic>;
  return json.keys.where((key) => !key.startsWith('@')).toSet();
}

String _messageValue(File file, String key) {
  final json = jsonDecode(file.readAsStringSync()) as Map<String, dynamic>;
  final value = json[key];
  return value is String ? value : '';
}
